utils/load_path.ml

```diff
@@ -34,12 +39,18 @@ module Dir = struct
      + treat [""] as the current directory. *)
   let readdir_compat dir =
     try
-      Sys.readdir (if dir = "" then Filename.current_dir_name else dir)
+      if dir = "" then
+        Dir, Sys.readdir Filename.current_dir_name
+      else if Sys.is_directory dir then
+        Dir, Sys.readdir dir
+      else
+        File, [|Filename.basename dir|]
```

Comment:
I'm not very familiar with the compiler code base, but where do you make sure that this file actually exists?

Reply:
Good point, fixed.

Reply:
Looks better, although I wish there was a stricter mode. In other words, the compiler would error if provided files that don't exist. This would be best for directories as well obviously, but there are some compatibility concerns. From the point of view of the build system, it would be more helpful if compiler is lax about this stuff.

Reply:
We expect the number of .cmi/.cmx files passed on the command-line to be potentially very large (transitive dependencies), especially if the feature is used to implement libraries in the build system.  So I'm slightly concerned by the extra cost induced by those `Sys.is_directory` and `Sys.file_exists`. Remember that we observed significant gains when we switch from file lookups (admittedly in may directories) to reading whole directories at once.

Cannot we not add any such calls?  Just remember which .cmi/.cmx file have been added (based on their extension, or specific CLI options...) to make them available if the compiler needs them.